File: packages/python/secrefs/providers/local.py
"""
Reads secrets from a gitignored, developer-local JSON file. Intended for
local development only.

Example `.secrefs.local.json`:
    { "mock-db": { "password": "hunter2", "user": "postgres" } }
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Dict, Optional, Union

from .base import ProviderHealth, SecretFetchRequest, SecretProvider, extract_field

DEFAULT_FILENAME = ".secrefs.local.json"


class LocalProvider(SecretProvider):
    name = "local"

    def __init__(
        self,
        file_path: Optional[Union[str, Path]] = None,
        cache_file: bool = False,
    ) -> None:
        """`cache_file` keeps the parsed file in memory instead of re-reading
        it per fetch. Off by default so an edit takes effect immediately -
        caching it meant editing .secrefs.local.json mid-session silently did
        nothing, which is the local-development shape of the same
        stale-secret problem ../ttl_cache.py exists to solve."""
        self._file_path = Path(
            file_path or os.environ.get("SECREFS_LOCAL_FILE") or (Path.cwd() / DEFAULT_FILENAME)
        )
        self._cache_file = cache_file
        self._cache: Optional[Dict[str, Any]] = None

    def _load(self) -> Dict[str, Any]:
        # The file is local and tiny, so re-reading it costs nothing worth
        # trading an edit-takes-effect guarantee for.
        if self._cache is not None and self._cache_file:
            return self._cache

        try:
            raw = self._file_path.read_text(encoding="utf-8")
        except OSError as exc:
            raise ValueError(
                f'[local] could not read local secrets file at "{self._file_path}": {exc}. '
                "This file is gitignored by convention - see .secrefs.local.json in .gitignore."
            ) from exc

        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ValueError(f'[local] "{self._file_path}" is not valid JSON: {exc}') from exc

        if not isinstance(parsed, dict):
            raise ValueError(f'[local] "{self._file_path}" must contain a top-level JSON object')

        self._cache = parsed
        return self._cache

    async def fetch_one(self, request: SecretFetchRequest) -> str:
        data = self._load()
        if request.path not in data:
            raise ValueError(f'[local] no entry for path "{request.path}" in {self._file_path}')

        entry = data[request.path]
        raw = entry if isinstance(entry, str) else json.dumps(entry)
        return extract_field(raw, request.field, provider=self.name, path=request.path)

    async def health_check(self) -> ProviderHealth:
        try:
            self._load()
            return ProviderHealth(provider=self.name, ok=True, message=str(self._file_path))
        except Exception as exc:  # noqa: BLE001 - surfaced to the caller as a health message
            return ProviderHealth(provider=self.name, ok=False, message=str(exc))

File: packages/node/src/providers/local.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import {
  BaseSecretProvider,
  type ProviderHealth,
  type SecretFetchRequest,
  extractField,
} from "./base.js";

const DEFAULT_FILENAME = ".secrefs.local.json";

export interface LocalProviderOptions {
  /** Overrides the file path. Defaults to $SECREFS_LOCAL_FILE or ./.secrefs.local.json */
  filePath?: string;
  /** Keep the parsed file in memory instead of re-reading per fetch.
   * Off by default so edits take effect immediately. */
  cacheFile?: boolean;
}

/**
 * Reads secrets from a gitignored, developer-local JSON file. Intended for
 * local development only - never point this at anything checked into
 * version control. Each top-level key is a secret path; its value is either
 * a plain string (returned as-is when no `#field` is requested) or an
 * object (JSON-stringified, then field-extracted as needed).
 *
 * Example `.secrefs.local.json`:
 * ```json
 * { "mock-db": { "password": "hunter2", "user": "postgres" } }
 * ```
 */
export class LocalProvider extends BaseSecretProvider {
  readonly name = "local";

  private readonly filePath: string;
  /** Re-read on every fetch. The file is local and tiny, and caching
   * it meant editing it mid-session silently did nothing. */
  private cache: Record<string, unknown> | null = null;
  private readonly cacheFile: boolean;

  constructor(options: LocalProviderOptions = {}) {
    super();
    this.filePath =
      options.filePath ??
      process.env.SECREFS_LOCAL_FILE ??
      path.join(process.cwd(), DEFAULT_FILENAME);
    this.cacheFile = options.cacheFile ?? false;
  }

  private async load(): Promise<Record<string, unknown>> {
    if (this.cache && this.cacheFile) return this.cache;

    let raw: string;
    try {
      raw = await readFile(this.filePath, "utf8");
    } catch (err) {
      throw new Error(
        `[local] could not read local secrets file at "${this.filePath}": ${
          err instanceof Error ? err.message : String(err)
        }. This file is gitignored by convention - see .secrefs.local.json in .gitignore.`,
      );
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(raw);
    } catch (err) {
      throw new Error(
        `[local] "${this.filePath}" is not valid JSON: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
      throw new Error(`[local] "${this.filePath}" must contain a top-level JSON object`);
    }

    this.cache = parsed as Record<string, unknown>;
    return this.cache;
  }

  async fetchOne(request: SecretFetchRequest): Promise<string> {
    const data = await this.load();
    if (!(request.path in data)) {
      throw new Error(`[local] no entry for path "${request.path}" in ${this.filePath}`);
    }

    const entry = data[request.path];
    const raw = typeof entry === "string" ? entry : JSON.stringify(entry);
    return extractField(raw, request.field, { provider: this.name, path: request.path });
  }

  async healthCheck(): Promise<ProviderHealth> {
    try {
      await this.load();
      return { provider: this.name, ok: true, message: this.filePath };
    } catch (err) {
      return {
        provider: this.name,
        ok: false,
        message: err instanceof Error ? err.message : String(err),
      };
    }
  }
}
